### PR TITLE
Add knowledge distillation for YOLO detection models

### LINKED ISSUE #24678: Yolo 26 Knowladge Distillation
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

I wanted to apply Knowledge Distillation (KD) to the YOLOv26m model using a YOLOv26x model as the teacher. Is there any documentation or code Implemention available for yolo 26 . I found implements for older yolo models can I use them on yolo 26 I found the following repos : 
- https://github.com/KefanZhan/YOLOv8-KD 
- https://github.com/danielsyahputra/yolo-distiller

### Additional

_No response_

### PR DESCRIPTION
**Pull request by lmycross**
## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
🚀 This PR adds full **knowledge distillation** support to Ultralytics training, plus documentation and tests, so smaller YOLO models can learn from larger teacher models for better accuracy without extra inference cost.

### 📊 Key Changes
- 🧠 Added a new **knowledge distillation training workflow** using the `distill_model` argument and `dis` loss weight.
- 🏗️ Introduced a new `DistillationModel` wrapper in `ultralytics/nn/distill_model.py` to manage:
  - frozen teacher models
  - trainable student models
  - feature extraction hooks
  - projector layers for feature alignment
  - combined standard + distillation loss
- 📚 Added a new **Knowledge Distillation guide** with:
  - quickstart examples for Python and CLI
  - recommended teacher/student model pairs
  - supported tasks: **detect, segment, pose, and obb**
  - performance tables for YOLO26 distilled models
  - FAQs, training tips, and resume examples
- 🧾 Added new training arguments to config/docs:
  - `distill_model`
  - `dis`
- 🔁 Added support for **resuming distillation training** from checkpoints, including rebuilding the teacher model automatically.
- 💾 Updated checkpoint/EMA saving so only the **student model** is exported, avoiding unnecessary teacher duplication and keeping inference artifacts lightweight.
- ⚙️ Updated trainer behavior to handle distillation safely, including:
  - disabling `compile` when distillation is enabled
  - freezing teacher parameters
  - logging an extra `dis_loss` during training
  - correctly applying class weights and validator settings to the student model
- 🧪 Added CLI and engine tests covering:
  - distillation training for detect, segment, pose, and obb
  - resume behavior
  - required YOLO26 teacher weights in test assets
- 🌐 Added docs navigation and homepage links so the new feature is easier to discover.

### 🎯 Purpose & Impact
- 🚀 Helps users train **smaller, faster YOLO26 models** that retain more of the accuracy of larger models.
- 📈 Offers a practical accuracy boost **without adding inference-time overhead**, which is especially useful for edge and real-time deployment.
- 🛠️ Makes distillation easier to use by integrating it directly into normal training flows instead of requiring custom code.
- 🔄 Improves reliability for longer experiments by supporting **resume-from-checkpoint** workflows.
- 📦 Keeps exported models efficient, since deployed artifacts contain only the student model.
- 📘 Lowers the adoption barrier with clear documentation, examples, and tested support across multiple tasks, though **detect** is the only task currently benchmarked for accuracy gains.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @lmycross, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This is an automated message to help with review readiness—an engineer will assist you shortly. Since this PR adds a new detection training feature with docs and tests, please review the checklist below to help ensure a smooth integration. 😊

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

File: ultralytics/nn/distill_model.py [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+            self.teacher_model.model[idx].register_forward_hook(FeatureHook(self._teacher_feats, idx))
+            self.student_model.model[idx].register_forward_hook(FeatureHook(self._student_feats, idx))
+
+        # Get feature dimensions via dummy forward pass (hooks capture outputs)
+        imgsz = student_model.args.imgsz
+        with smart_inference_mode():
+            teacher_model(torch.zeros(2, 3, imgsz, imgsz).to(device))
+            student_model(torch.zeros(2, 3, imgsz, imgsz).to(device))
+        teacher_output = [self._teacher_feats[idx] for idx in self.feats_idx]
+        student_output = [self._student_feats[idx] for idx in self.feats_idx]
```
Comment (UltralyticsAssistant) [inline ultralytics/nn/distill_model.py:93]: ⚠️ **HIGH**: This shape-probing forward runs the student in its current mode, which is train mode for freshly built models. Any BatchNorm layers will update their running stats from an all-zero tensor before the first real batch, subtly perturbing training and resume behavior. Please switch the student to eval for this probe and restore its previous mode afterward.

File: ultralytics/utils/torch_utils.py [hunk 2]
```diff
@@ -691,6 +693,9 @@ def update(self, model):
 
             msd = unwrap_model(model).state_dict()  # model state_dict
             for k, v in self.ema.state_dict().items():
+                if "teacher" in k:
```
Comment (UltralyticsAssistant) [inline ultralytics/utils/torch_utils.py]: ⚠️ **HIGH**: Skipping teacher keys during EMA updates avoids changing the frozen teacher, but it doesn't avoid the bigger cost: `ModelEMA(self.model)` has already deep-copied the entire teacher into the EMA model. In distillation runs that means carrying an extra full teacher copy for the whole training job, which is a large memory/checkpoint overhead. The distillation path should build EMA from the student only, or strip the teacher from the EMA model up front.

File: ultralytics/nn/distill_model.py [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+        Args:
+            teacher_model (str | nn.Module): Teacher model checkpoint path or module.
+            student_model (nn.Module): Student model module to be trained.
+        """
+        super().__init__()
+        if isinstance(teacher_model, str):
+            teacher_model = load_checkpoint(teacher_model)[0]
+        device = next(student_model.parameters()).device
+        self.teacher_model = teacher_model.to(device)
+        self._freeze_teacher()
```
Comment (UltralyticsAssistant) [inline ultralytics/nn/distill_model.py:74]: 💡 **MEDIUM**: `distill_model` is exposed as a model path, but this only loads the teacher when `teacher_model` is a `str`. Passing a `Path` from Python will fall through to `teacher_model.to(device)` and raise. Please accept path-like inputs here as well so the new API behaves like the rest of the loading paths.

File: ultralytics/nn/distill_model.py [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+
+        # Hook-based feature capture: identical for teacher and student
+        self._teacher_feats = {}
+        self._student_feats = {}
+        for idx in self.feats_idx:
+            self.teacher_model.model[idx].register_forward_hook(FeatureHook(self._teacher_feats, idx))
+            self.student_model.model[idx].register_forward_hook(FeatureHook(self._student_feats, idx))
+
+        # Get feature dimensions via dummy forward pass (hooks capture outputs)
+        imgsz = student_model.args.imgsz
```
Comment (UltralyticsAssistant) [inline ultralytics/nn/distill_model.py:86]: 💡 **MEDIUM**: The hooks are registered unconditionally here, but resumed checkpoints already reattach distillation hooks in `__setstate__`. Re-wrapping a resumed teacher will stack another set of hooks, so each teacher forward does extra work and keeps stale feature dicts alive. It would be safer to track/remove your own hook handles before registering new ones.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Nice feature addition overall, but I found a few implementation issues worth fixing before merge: the dummy forward used for shape inference mutates student BN stats, path-like teacher inputs can currently crash, resumed runs can accumulate duplicate hooks, and the EMA path still retains a full frozen teacher copy.

💬 Posted 4 inline comments

**Comment by Laughing-q** [comment]
## ⚡ Actions Trigger

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

GitHub Actions below triggered via workflow dispatch for this PR at 2026-06-23 12:57:58 UTC with `@ultralytics/run-ci` command
(available commands are `@ultralytics/run-all`, `@ultralytics/run-ci`, and `@ultralytics/run-docker`):

* ✅ [CI](https://github.com/ultralytics/ultralytics/actions/runs/28027853281): `ci.yml` run 24118

**Comment by Laughing-q** [comment]
@glenn-jocher this is ready from my side. also added a hint at the home page to show the latest knowledge distillation feature
pic-260623-2258-24

**Review by Laughing-q — APPROVED** [review]

**Comment by Laughing-q** [comment]
> @glenn-jocher this is ready from my side. also added a hint at the home page to show the latest knowledge distillation feature pic-260623-2258-24

@onuralpszr BTW I think probably we should do the same to tracker as well

**Comment by glenn-jocher** [comment]
## ⚡ Actions Trigger

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

GitHub Actions below triggered via workflow dispatch for this PR at 2026-06-25 02:08:50 UTC with `@ultralytics/run-all` command
(available commands are `@ultralytics/run-all`, `@ultralytics/run-ci`, and `@ultralytics/run-docker`):

* ✅ [CI](https://github.com/ultralytics/ultralytics/actions/runs/28142139134): `ci.yml` run 24213
* ✅ [Publish Docker Images](https://github.com/ultralytics/ultralytics/actions/runs/28142139777): `docker.yml` run 2412

**Comment by UltralyticsAssistant** [comment]
Merged! 🎉

Huge thanks to @lmycross for driving this forward, with valuable contributions from @onuralpszr, @glenn-jocher, and @Laughing-q.

> “Great things are done by a series of small things brought together.” — Vincent van Gogh

This PR is a great example of that idea in action: training workflow updates, a dedicated distillation wrapper, docs, resume support, lightweight student-only exports, and multi-task test coverage all come together to make knowledge distillation a practical and polished part of Ultralytics. The result is a meaningful step toward helping users train smaller YOLO26 models with stronger accuracy and no extra inference cost.

Really appreciate the thoughtful implementation and collaboration here 👏

**Comment by glenn-jocher** [comment]
Final review and hardening before merge. Summary of changes layered on the original implementation:

**Bug fixes**
- `compile=True` + distillation crash (`AttributeError` on `.stride`): the student was compiled before being wrapped, so attribute copying failed. Compilation is now disabled when distilling, which also restores DDP `find_unused_parameters` handling for the frozen teacher.
- `multi_scale` training crash in `torch.split`: teacher score split sizes are now derived per-batch from the live neck features instead of cached at init. Added `multi_scale` coverage to the resume test.
- `cls_pw` class weights silently dropped under distillation: weights are now applied to the student model that builds the loss criterion.

**Simplifications (Delete > Replace > Add)**
- Reverted the `smart_inference_mode` rewrite; the distillation shape-probe uses `torch.no_grad()` directly, leaving the shared utility and its ~30 call sites untouched.
- Removed dead code (always-true assert, unreachable projection guard, unused `fuse()`, impossible-state `hasattr` guard).
- Registered `dis` in `CFG_FLOAT_KEYS` for type validation, corrected the `DistillationModel` docstring example, and reused the validator's predictions in the validation-time distillation loss.

**Verified**
- Trains for detect/segment/pose/obb, survives save→resume, and strips to a clean student `DetectionModel` (no teacher/projector) — student-only export confirmed.
- Full fast CI and slow CI (`@ultralytics/run-all`) green; version bumped to 8.4.77.
